docs: add ADR 063: core API ADR + update app wiring ADR

aaronc opened the pull request:
<!--
The default pull request template is for types feat, fix, or refactor.
For other templates, add one of the following parameters to the url:
- template=docs.md
- template=other.md
-->

## Description

Closes #12923 

This PR adds ADR 063 which describes the core module API in #12239.

It also updates ADR 057: App Wiring to reflect integration with this ADR. ADR 063 is effectively the part 2 of app wiring mentioned in ADR 057.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

marbar3778 commented on docs/architecture/adr-061-core-module-api.md:
```diff
@@ -0,0 +1,490 @@
+* the core API doesn't implement *any* functionality, it just defines types
+* go stable API management principles are followed (TODO: link)
+
+Runtime modules which implement the core API are *intentionally* separate from the core API in order to enable more 
```
will this work if a user does not want to use runtime/DI?

aaronc replied:
It will work without DI. If someone wants to do manual wiring, codegen requires exported providers so manual app.go is totally possible.

Runtime basically encapsulates what is now BaseApp + module manager, and will extend these and eventually replace the current implementations (what we've called BaseApp 2.0 with ABCI++, etc). I'll add a bit more text to the ADR defining this.

So runtime will be necessary, but it will be possible to wire it manually, or use DI, or DI codegen which is like manual wiring that the framework does for you.

alexanderbez commented on docs/architecture/adr-061-core-module-api.md:
```diff
@@ -0,0 +1,490 @@
+
+## Abstract
+
+A new core API is proposed to replace the existing `AppModule` and `sdk.Context` frameworks a set of core services
+and a new `Handler` struct. This core API aims to:
+- be simpler,
+- more extensible
+- and more stable than the current framework
+while also
+- enabling deterministic events and queries, and
+- supporting event listeners
+- and [ADR 033: Protobuf-based Inter-Module Communication](./adr-033-protobuf-inter-module-comm.md) clients.
```
```suggestion
A new core API is proposed to replace the existing `AppModule` and `sdk.Context` frameworks, a set of core services, and a new `Handler` struct. This core API aims to:

* Be simpler
* More extensible
* More stable than the current framework
* Enable deterministic events and queries
* Support event listeners and [ADR 033: Protobuf-based Inter-Module Communication](./adr-033-protobuf-inter-module-comm.md) clients
```

alexanderbez commented on docs/architecture/adr-061-core-module-api.md:
```diff
@@ -0,0 +1,490 @@
+
+## Context
+
+Historically modules have exposed their functionality to the state machine via the `AppModule` and `AppModuleBasic`
+interfaces which have the following shortcomings:
```
I don't know what this means? I recon you could argue they expose their functionality via Keepers. The `AppModule*` interfaces are really more app wiring UX primitives.

aaronc replied:
Maybe "state machine" isn't the right term. They expose their functionality to the SDK framework/BaseApp with these interfaces

alexanderbez commented on docs/architecture/adr-061-core-module-api.md:
```diff
@@ -0,0 +1,490 @@
+
+Historically modules have exposed their functionality to the state machine via the `AppModule` and `AppModuleBasic`
+interfaces which have the following shortcomings:
+* both `AppModule` and `AppModuleBasic` need to be defined which is counter-intuitive
```
WDYM?

kocubinski replied:
The distinction between features on `AppModule` and `AppModuleBasic` seems arbitrary.   So does the use of the word "basic".

alexanderbez commented on docs/architecture/adr-061-core-module-api.md:
```diff
@@ -0,0 +1,490 @@
+Historically modules have exposed their functionality to the state machine via the `AppModule` and `AppModuleBasic`
+interfaces which have the following shortcomings:
+* both `AppModule` and `AppModuleBasic` need to be defined which is counter-intuitive
+* apps need to implement the full interfaces, even parts they don't need
```
Not technically, as they can define no-ops for many of them, which they do.

kocubinski replied:
Perhaps then the con is: "Modules are required to supply unnecessary boilerplate to implement no-op"

kocubinski replied:
Although, I _think_ some of this could be addressed with breaking refactors to `AppModule`, `AppModuleBasic`.  That is, if [AppModuleBasic](https://github.com/cosmos/cosmos-sdk/blob/249643325182a85dc60d2baae7d867b1db765467/types/module/module.go#L48-L60) was decomposed into domain-specific extension interfaces.

alexanderbez commented on docs/architecture/adr-061-core-module-api.md:
```diff
@@ -0,0 +1,490 @@
+interfaces which have the following shortcomings:
+* both `AppModule` and `AppModuleBasic` need to be defined which is counter-intuitive
+* apps need to implement the full interfaces, even parts they don't need
+* interface methods depend heavily on unstable third party dependencies, in particular Tendermint
```
This isn't a con IMO

aaronc replied:
Depending on particular versions of Tendermint means you can't have a module which simultaneously supports multiple SDK versions in a hypothetical world where ABCI++ exists but maybe some users are more conservative about adopting it. More generally, it means the ecosystem needs to move in fits and starts - all modules must advance together, rather than allowing a "matrix" of compatible versions. I consider this a big con. Will try to explain this a bit more in the ADR

alexanderbez commented:
This is going to take me a while to review. I would love as much review as possible from those that this actually impacts the most -- app devs, client and tooling devs, etc...

kocubinski commented on docs/architecture/adr-061-core-module-api.md:
```diff
@@ -0,0 +1,490 @@
+will allow modules to send messages to and make queries against other modules as described in [ADR 033](./adr-033-protobuf-inter-module-comm.md).
+
+A single `grpc.ClientConnInterface` will be used for both MsgClient's and QueryClient's and the state machine will
+make the presence or absence of the protobuf option `cosmos.msg.v1.service` which will be required for all
```
>and the state machine will make the presence or absence of the protobuf option cosmos.msg.v1.service which will be required for all valid Msg service types.

I don't understand this sentence, is there a typo or something?

aaronc replied:
Yes, I'll fix

kocubinski commented on docs/architecture/adr-061-core-module-api.md:
```diff
@@ -0,0 +1,490 @@
+
+### Backwards Compatibility
+
+Early versions of runtime modules should aim to support as much as possible modules built with the existing
```
The wording here is runtime _modules_ but in practice the SDK team will implement just one, at least initially, is that right?

I'm having a hard time visualizing how runtime would straddle the current `AppModule / sdk.Context` paradigm and   the proposed composition through `Handler`.

aaronc replied:
Well, I think we would implement one for the existing tendermint version, and another one once ABCI++ comes around possibly deprecating `AppModule` altogether. The legacy version might have a fairly long life span of patch updates.

As for the second part, you would just have some sort of module manager which can deal with both `AppModule` and `Handler` appropriately.

kocubinski commented on docs/architecture/adr-061-core-module-api.md:
```diff
@@ -0,0 +1,490 @@
+* go stable API management principles are followed (TODO: link)
+
+Runtime modules which implement the core API are *intentionally* separate from the core API in order to enable more 
+forks of the runtime module than is possible with the SDK's current tightly coupled `BaseApp` design while still
```
What could a fork of SDK's default runtime look like in practice?  Is this a code fork or is runtime to be composable so that users of the sdk can add/adjust functionality?  Is there a need for that currently (is this why osmosis forked)?

kocubinski commented:
Is this more or less right? 

#### Presently

`sdk.Context` is a combination of Service Locator (e.g. `.KvStore`, `GasMeter`, `EventManager`) and runtime state (`txBytes`, `checkTx`) repository.  As we extend SDK functionality `sdk.Context` will necessarily grow to include flags, data and services to support the added features.  All module functions which take an `sdk.Context` have access to the entirety of features provided by it, this includes read/write access to the entire state store.

Module behavior is defined at *compile time* by implementing `AppModule*` interfaces.  This requires as lot of boilerplate when implementing a new module.

#### Proposed

Module functions do not receive an `sdk.Context` and instead register their required view (i.e. dependencies) of application context at *compile time* via `depinject`.  This will require lots of refactoring to accommodate constructor injection of required services.  The runtime module provides the implementation to fetch or construct these services.  This limits the blast of radius of a single module and ensures that they all interact with common functionality via a single well-defined Core API.

Module behavior is defined *at runtime* (app startup) by interacting with `Handler` and setting behavior functions equivalent to the current set in `AppModule*`.

aaronc commented:
> Is this more or less right?

Sounds pretty accurate @kocubinski .

> Update: https://github.com/cosmos/cosmos-sdk/compare/aaronc/core-api...kocubinski/core-api-v2 shows a rough draft with module capabilities fleshed out a bit more.

This is an interesting alternative approach @kocubinski. I'll need to reflect on it a bit. Seems similar to expected keepers. The big downside to this alternate approach that I see is that it doesn't seem possible to know whether a service is available until block processing actually starts whereas with the depinject approach you know at initialization or compile time (with codegen). Does that sound accurate?

aaronc commented:
> This is going to take me a while to review. I would love as much review as possible from those that this actually impacts the most -- app devs, client and tooling devs, etc...

Agreed, I'm mainly wanting core devs to have some shared context first to address any obvious concerns.

aaronc commented:
I've made some updates based on comments here and marked this R4R

aaronc commented:
I have made a few updates here to simplify the API with a "bundled" `appmodule.Service` so that there is a single service that can be used as a dependency.

I also made some updates to the ADR-033 interfaces.

This definitely needs more reviews

aaronc commented on docs/architecture/adr-061-core-module-api.md:
````diff
@@ -0,0 +1,562 @@
+The `Handler` struct type will implement `grpc.ServiceRegistrar` interface which will allow for registering
+`MsgServer` and `QueryServer` implementations directly with the `Handler` struct like this:
+```go
+    h := &appmodule.Handler{}
+	types.RegisterMsgServer(h, keeper.NewMsgServerImpl(am.keeper))
+	types.RegisterQueryServer(h, am.keeper)
````
What extension interface would you propose for registering services, events listeners and upgrade handlers? I remember some people not like the current Configurator approach. 

The advantage of a struct is that it's explicit about what is supported whereas with extension interfaces you're sort of guessing about whether some other component will type cast and handle it.

amaurym replied:
I was thinking a bunch of small interfaces like

```go
package appmodule

type HasMsgServer interface {
  RegisterMsgService(s grpc.Server)
}

type HasGrpcServer interface {
  RegisterGrpcService(s grpc.Server)
}

type HasMigrations interface {
  RegisterMigrations() map[uint]MigrationHandler // `map` is maybe not ideal, maybe a more well-defined struct as return type
}

type HasEventListeners interface {
  RegisterEventListeners() map[string]EventHandler // string == proto full name of typed event
}
```

Then, at a glance we can know what a module does:

```go
// in x/{...}/module.go
var (
  appmodule.AppModule      = AppModule{}
  appmodule.HasMsgServer   = AppModule{}
  appmodule.HasGrpcServer  = AppModule{}
  appmodule.HasMigrations  = AppModule{}
)
```

amaurym replied:
> The advantage of a struct is that it's explicit about what is supported whereas with extension interfaces you're sort of guessing about whether some other component will type cast and handle it.

True, there would need to be good docs on all available extension interfaces if we go this way. On the other hand, it's less invasive IMO compared to the status quo.

aaronc replied:
I don't really have a strong preference. I'd like a good enough design we can agree on

aaronc replied:
If we do extension interfaces, I agree that minimal is better

marbar3778 commented on docs/architecture/adr-061-core-module-api.md:
```diff
@@ -0,0 +1,562 @@
+
+## Abstract
+
+A new core API is proposed to replace the existing `AppModule` and `sdk.Context` frameworks a set of core services
```
```suggestion
A new core API is proposed to be added as a way to develop cosmos-sdk applications that will in the future replace the existing `AppModule` and `sdk.Context` frameworks a set of core services
```
?

marbar3778 commented on docs/architecture/adr-061-core-module-api.md:
```diff
@@ -0,0 +1,562 @@
+    ChainID string
+    Height int64
+    Time time.Time
+    Hash []byte
```
what hash is this?

I think things like IBC need other things like validatorhash and next validatorhashes.

aaronc replied:
I was thinking last commit hash, but since there are multiple hashes maybe best not to include one here. IBC would probably just depend on the specific version of tm it's compatible with... unless that sort of stuff is expected to be stable enough between tm versions

marbar3778 replied:
I think it should be fine to include all the hashes here since its in line with the naming of the struct. If we opt not to can we add a few sentences on how to get this info

marbar3778 commented on docs/architecture/adr-061-core-module-api.md:
```diff
@@ -0,0 +1,562 @@
+
+The genesis `Handler` functions - `DefaultGenesis`, `ValidateGenesis`, `InitGenesis` and `ExportGenesis` - are specified
+against the `GenesisSource` and `GenesisTarget` interfaces which will abstract over genesis sources which may be single
+JSON files or collections of JSON objects that can be efficiently streamed. They will also include helper functions for
```
why both proto and json? why not only proto to align with proto use in the repo?

aaronc replied:
What about cases where things can't be modeled as single proto objects or if people want to use different encodings for some reason? IMHO better to be more permissive, but make it as convenient as possible.

marbar3778 commented on docs/architecture/adr-061-core-module-api.md:
````diff
@@ -0,0 +1,562 @@
+
+```go
+type GenesisSource interface {
+	ReadMessage(proto.Message) error
+	OpenReader(field string) (io.ReadCloser, error)
+	ReadRawJSON() (json.RawMessage, error)
````
we should opt to use one and abstract the other, this is more boiler plate for users

aaronc replied:
Why is it more boilerplate? The user doesn't implement this interface, they just use it how they like.

marbar3778 commented on docs/architecture/adr-061-core-module-api.md:
```diff
@@ -0,0 +1,562 @@
+
+#### Event Listeners
+
+Handlers allow event listeners for typed events to be registered that will be called deterministically whenever those
```
what about legacy events?

aaronc replied:
This design doesn't allow for listening to legacy events, just typed events. My reasoning is that those events weren't formalized in a way to make them relevant to state machine consumers, so we should assume they're intended for clients only

amaurym replied:
I also agree limiting state-machine events to only typed events is desirable. It can also be seen as an extra motivation to migrate to typed events asap (pending perf fixes).

marbar3778 replied:
will runtime still support legacy events?

aaronc replied:
Yes, there is an `EmitLegacy` method so that these types of events are supported for perpetuity

marbar3778 commented on docs/architecture/adr-061-core-module-api.md:
````diff
@@ -0,0 +1,562 @@
+}
+```
+
+In the [ADR 057: App Wiring](./adr-057-app-wiring-1.md) it is specified that each version of each module should be
````
can you provide an example of non DI way of doing this.

aaronc replied:
Will get to that... It will be easier once codegen is wrapped up because there are some API changes needed to our current setup

kocubinski commented on docs/architecture/adr-061-core-module-api.md:
```diff
@@ -0,0 +1,562 @@
+
+#### `appmodule.Service` Bundle Service
+
+To allow modules to declare a dependency on a single "bundle" service, runtime modules should provide an implementation
```
Is the intent with this bundled service to make adoption easier?  In the case of manual wiring I'd need to construct one service (constructor defined in `runtime`) for each module is that right?  Might be even more boilerplate than without this uber service.

aaronc replied:
Yes, the intent is to make adoption easier. Not sure it'd be a lot more boilerplate. Depends on how you construct the service, but shouldn't be harder than constructing store keys

kocubinski commented on docs/architecture/adr-061-core-module-api.md:
````diff
@@ -0,0 +1,562 @@
+}
+```
+To maintain API compatibility, if new core services are added, a new `cosmossdk.io/coreappmodule/v2.Service` should
+be added which extends this service and bundles the new core services, ex:
````
Historically how often has this happened? Seems like a big downside of this approach.

aaronc replied:
Not really sure, but why is it a big downside? We just bump the version

kocubinski replied:
As a user of the Cosmos SDK it would make upgrading to a new core API version more of a refactor and therefore more cumbersome, no?

aaronc replied:
Why? They just need to change the import

kocubinski replied:
Changing a `go.mod` import is less work than changing a `go.mod` import and import statements across `.go.` files, which is required since the package is now `v2`.  Maybe "big" drawback should be modified to "minor" drawback.

aaronc replied:
Why do you need to change so many imports? You only need to change imports that require the new service

kocubinski replied:
On reflection I guess it's no more work than updating a method signature to support a new service.  In fact it's probably easier to change an import statement than a func sig.  Am I thinking about this right?

aaronc replied:
I think so

amaurym commented on docs/architecture/adr-061-core-module-api.md:
````diff
@@ -0,0 +1,562 @@
+Modules can use these services like this:
+```go
+func (k msgServer) Send(ctx context.Context, msg *types.MsgSend) (*types.MsgSendResponse, error) {
+	store := k.kvStoreSvc.OpenKVStore(ctx)
````
related to the above comment: it seems that we always do `k.{someService}.{whatWeActuallyWant}`. Most services only have 1 method. This new `.Open` seems unnecessary here.

For some of these core services: can we simplify it to simply `k.{whatWeActuallyWant}`.

```diff
type keeper struct {
	depinject.In

-	KVStoreKey       store.KVStoreService
+	KVStoreKey       store.KVStore
	BlockInfoService blockinfo.Service
-	EventService     event.Service
+	EventMgr         event.Manager
	GasService       gas.Service
}
```

aaronc replied:
We still need to interact with the context to get a valid instance. Maybe the bundle service simplifies this?

amaurym replied:
Ah I see. I remember a discussion about adding the `context.Context` argument to all Get/Set/Delete/etc methods. Maybe Open is better.

aaronc replied:
If there is more than one store operation in a method than open will be shorter

amaurym commented on docs/architecture/adr-061-core-module-api.md:
```diff
@@ -0,0 +1,562 @@
+	GetMeter(context.Context) Meter
+	GetBlockMeter(context.Context) Meter
+	WithMeter(ctx context.Context, meter Meter) context.Context
+	WithBlockMeter(ctx context.Context, meter Meter) context.Context
```
IMO `WithBlockMeter` should not be in the default Gas service. It shouldn't be exposed to most module developers. Probably only runtime and maybe antehandlers should deal with BlockGas.

aaronc replied:
Yeah it's worth having a discussion about. IMHO there's a lot of unsafe stuff on the current sdk.Context

amaurym commented on docs/architecture/adr-061-core-module-api.md:
```diff
@@ -0,0 +1,562 @@
+manual wiring. The individual services described below are all bundled in a convenient `appmodule.Service`
+"bundle service" so that for simplicity modules can declare a dependency on a single service.
+
+#### Store Services
```
Could you give an example of how one would do `ctx.CacheContext()` with this store service? It doesn't seem clear to me right now

aaronc replied:
That's maybe something that needs to be added

aaronc replied:
I think there are two ways to do this:
1. any service which can operate in cache mode offers its own API for this, i.e. `KVStore.CacheWrap`, `EventService.CacheWrap`
2. there is some way that caching happens across multiple services which is how `CacheContext` behaves now

The downside to 1, is that you need to remember to cache all 3 store types and it isn't obvious how to do this across inter-module calls. So I think we need some way to do 2, probably with a `CacheService` that has a way of notifying other services they need to cache:

```go
type CacheService interface {
  CacheContext(context.Context) context.Context)
}
```

amaurym commented on docs/architecture/adr-061-core-module-api.md:
```diff
@@ -0,0 +1,562 @@
+
+The design principles of the core API are as follows:
+* everything that a module wants to interact with in the state machine is a service
+* all services coordinate state via `context.Context` and don't try to recreate the "bag of variables" approach of `sdk.Context`
```
Just to understand this line: this "bag of variables" still exists, and will always exist with this new design, correct? It's just that we try our best to hide it to module developers.

As long as most keeper methods (e.g. msg server methods) have access to a `context.Context`, they can always peek into it to see all available variables, even those not exposed by said keeper's available services.

aaronc replied:
Sort of, but there's no way to do a range operation over everything in a context. You need to hold a key object which can be non-forgeable.

amaurym replied:
But you can still access `sdk.Context` via the unwrap method.

Is the plan to deprecate `sdk.Context`? If yes, it's worth mentioning this in the ADR, and start adding `// Deprecated` comments on it as soon as this approach is approved by the community.

aaronc replied:
For now `sdk.Context` can still be accessed via unwrap. In my mind deprecating `sdk.Context` is somewhat orthogonal because that's a runtime impl detail, but I can say something here

amaurym reviewed (commented):
Overall I like that this approach allows each keeper to define the dependencies/services it needs 👍 , and those dependencies need not be provided by the SDK. So concept ACK, but have some final questions before approving.

aaronc commented on docs/architecture/adr-061-core-module-api.md:
```diff
@@ -0,0 +1,562 @@
+	// Emit emits events to both clients and state machine listeners. These events MUST be emitted deterministically
+	// and should be assumed to be part of blockchain consensus.
+    Emit(proto.Message) error
+	
```
The reason would be if we want to allow overriding the event manager like the current SDK context

julienrbrt commented:
@aaronc Can we split this PR and land the update of the App Wiring ADR separately as this will already be included in v0.47.
Otherwise, the docs link to an outdated ADR :)

aaronc commented:
> @aaronc Can we split this PR and land the update of the app wiring ADR separately (without the core API stuff) as this will already be included in v0.47.
> Otherwise, the docs link to an outdated ADR :)

See https://github.com/cosmos/cosmos-sdk/pull/13736. Going to put this into draft and make that one R4R

aaronc commented:
Reopening this because it would be great to at least merge this as `DRAFT`, if not `PROPOSED`. Updating the test to reflect the extension interface approach before marking R4R

tac0turtle commented on docs/architecture/adr-063-core-module-api.md:
```diff
@@ -0,0 +1,546 @@
+a version of the module which can be run with 2 or 3 different versions of the SDK or 2 or 3 different versions of
+another module. This lock-step coupling slows down overall development within the ecosystem and causes updates to
+components to be delayed longer than they would if things were more stable and loosely coupled.
+
```
can we add an alternative section to this? its kind of hard to gauge design when the decision is the only consideration

tac0turtle commented on docs/architecture/adr-063-core-module-api.md:
```diff
@@ -0,0 +1,546 @@
+* everything that a module wants to interact with in the state machine is a service
+* all services coordinate state via `context.Context` and don't try to recreate the "bag of variables" approach of `sdk.Context`
+* all independent services are isolated in independent packages with minimal APIs and minimal dependencies
+* the core API should be minimalistic and designed for long-term support (LTS)
```
how do we guarantee the proposed design is the right one and wont need to make changes?

tac0turtle commented on docs/architecture/adr-063-core-module-api.md:
```diff
@@ -0,0 +1,546 @@
+type Manager interface {
+    // Emit emits events to both clients and state machine listeners. These events MUST be emitted deterministically
+    // and should be assumed to be part of blockchain consensus.
+    Emit(proto.Message) error
```
why does this have to be a proto.message instead of a golang type?

aaronc replied:
Because this supports typed events as the first class type of events. We can't just use a regular golang type as a typed event.

tac0turtle replied:
rereading the typed events adr I'm not convinced this should be protobuf. is there reasoning proto was decided on here?

tac0turtle commented on docs/architecture/adr-063-core-module-api.md:
````diff
@@ -0,0 +1,546 @@
+```
+
+Typed events emitted with `Emit` can be observed by other modules and thus must be deterministic and should be assumed
+to be part of blockchain consensus (whether they are part of the block or app hash is left to the runtime to specify).
````
how would runtime do this? would it be up to the application developer?

aaronc replied:
There is a way for registering event listeners described later.

tac0turtle commented on docs/architecture/adr-063-core-module-api.md:
```diff
@@ -0,0 +1,546 @@
+Design questions:
+* should we allow overriding event managers (like `sdk.Context.WithEventManager`)?
+
+#### Block Info Service
```
I think this service may be better named consensus as it has to do with consensus related things. 

Secondly, can you briefly comment on how this interface was chosen? what stake holders were consulted for the design, what use cases were considered? to me it seems we are missing stuff here in order to support different use cases. The sdk itself is not a good tell on what is used vs not

tac0turtle commented on docs/architecture/adr-063-core-module-api.md:
````diff
@@ -0,0 +1,546 @@
+}
+```
+
+Only a limited set of modules need any other information from the Tendermint block header and specific versions
````
if we are thinking of abstracting this, I would propose we spend some more time thinking through this in order to see if we could abstract tendermint away from the Cosmos SDK in order to make it agnostic to consensus. I have some ideas on what could be done here.

tac0turtle commented on docs/architecture/adr-063-core-module-api.md:
```diff
@@ -0,0 +1,546 @@
+    ChainID string
+    Height int64
+    Time time.Time
+    Hash []byte
```
still don't know what hash this is? 

lastresulthash, app hash, last validator hash?

tac0turtle commented on docs/architecture/adr-063-core-module-api.md:
````diff
@@ -0,0 +1,546 @@
+
+In order for `BeginBlock`, `EndBlock` and `InitGenesis` to send back validator updates and retrieve full Tendermint
+block headers, the runtime module for a specific version of Tendermint could provide services like this:
+```go
+type ValidatorUpdateService interface {
+    SetValidatorUpdates(context.Context, []abci.ValidatorUpdate)
+}
+
+type BeginBlockService interface {
+    GetBeginBlockRequest(context.Context) abci.RequestBeginBlock 
+}
+```
````
I don't think we should allow modules to deal with anything tendermint related. Tendermint is a single consensus protocol that the module shouldn't care about. Ideally sdk defines an interface or concrete type that base app/abci turns into a tendermint type. This way modules don't need to depend on tendermint when they are spun out.

aaronc replied:
Is it fine if x/staking depends on tendermint? What I'm proposing here is that only x/staking would need that dependency and other modules wouldn't. We could also define some other validator update type in core or runtime too.

tac0turtle replied:
Yes, this is fine. In the future when we decide to own the abci types ourselves and have a transition Layer for tendermint we can modify it. For now this isnt a worry nor should it block anything

tac0turtle commented on docs/architecture/adr-063-core-module-api.md:
```diff
@@ -0,0 +1,546 @@
+#### Event Listeners
+
+Handlers allow event listeners for typed events to be registered that will be called deterministically whenever those
+events are called in the state machine. Event listeners are functions that take a context, the protobuf message type of
```
why does this need to be a proto message?

aaronc replied:
Again because of ADR 032 typed events

tac0turtle commented on docs/architecture/adr-063-core-module-api.md:
```diff
@@ -0,0 +1,546 @@
+Event listeners provide a standardized alternative to module hooks, such as `StakingHooks`, even though these hooks
+are still supported and described in [ADR 057: App Wiring](./adr-057-app-wiring-1.md)
+
+Only one event listener per event type can be defined per module. Event listeners will be called in a deterministic
```
I believe these events are different than client events, if so we should make sure to identify that. This could get confusing

aaronc replied:
These events are both client and internal events. The legacy "untyped" events are client only.

tac0turtle replied:
oh didnt know this. Is this mentioned in another ADR?do these events get added to consensus if they are used internally?

aaronc replied:
Yes. This is mentioned first in this ADR I think. Maybe it's too big of a scope and we should save it for a separate ADR

aaronc commented:
I've updated this ADR to _only_ include stuff that has already been merged into core. I have removed gas, block info, and upgrades which should probably be part of this ADR but in an update; event listeners which should probably be a separate ADR; and ADR 033 stuff which should probably just go in an update of that ADR.

Is this good to merge now?

tac0turtle reviewed (approved):
thank you for making the changes. Super excited to see this come together

tac0turtle commented:
Marking this to be merged since its already mostly implemented, which means we have had consensus. No need to block the pr
